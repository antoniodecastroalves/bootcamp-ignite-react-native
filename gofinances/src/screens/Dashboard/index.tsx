import React, { useState, useEffect, useCallback } from 'react'
import { ActivityIndicator } from 'react-native'
import AsyncStorage from '@react-native-async-storage/async-storage'

import { useFocusEffect  } from '@react-navigation/native'
import { useTheme } from 'styled-components';

import { Card } from '../../components/Card/index';
import { TransactionCard, TransactionCardProps } from '../../components/TransactionCard';
import { 
  Container,
  Header,
  UserWrapper,
  UserInfo,
  Photo,
  User,
  UserGreeting,
  UserName,
  Icon,
  CardsContainer,
  Transactions,
  TransactionList,
  Title,
  LogoutButton,
  LoadContainer,
} from './styles'

export interface DataListProps extends TransactionCardProps {
  id: string;
}

interface CardProps {
  amount: string;
}

interface CardData {
  income: CardProps;
  outcome: CardProps;
  total: CardProps;
}

export function Dashboard() {
  const theme = useTheme()
  const [isLoading, setLoading] = useState(true)
  const [transactions, setTransactions] = useState<DataListProps[]>([]);
  const [cardData, setCardData] = useState<CardData>({} as CardData);

  async function loadTransactions() {
    const dataKey = '@gofinances:transactions'
    const response = await AsyncStorage.getItem(dataKey);
    const currentTransactions = response ? JSON.parse(response) : [];

    let incomeTotal = 0;
    let outcomeTotal = 0;

    const formattedTransactions: DataListProps[] = currentTransactions.map((transaction: DataListProps) => {

      if (transaction.type === 'income') {
        incomeTotal += Number(transaction.amount);
      } else {
        outcomeTotal += Number(transaction.amount);
      }

      const amount = Number(transaction.amount)
        .toLocaleString('pt-BR', { style: 'currency', currency: 'BRL' });

      const formattedDate = Intl.DateTimeFormat('pt-BR', {
        day: '2-digit',
        month: '2-digit',
        year: 'numeric',
      }).format(new Date(transaction.date));

      return {
        ...transaction,
        amount,
        date: formattedDate,
      }
    })

    setTransactions(formattedTransactions);
    setCardData({
      income: {
        amount: incomeTotal.toLocaleString('pt-BR', { 
          style: 'currency', 
          currency: 'BRL',
        }),
      },
      outcome: {
        amount: outcomeTotal.toLocaleString('pt-BR', { 
          style: 'currency', 
          currency: 'BRL',
        }),
      },
      total: {
        amount: (incomeTotal - outcomeTotal).toLocaleString('pt-BR', {
          style: 'currency',
          currency: 'BRL',
        }),
      },
    })

    setLoading(false)
  }

  useEffect(() => {
    loadTransactions();
  },[]);

  useFocusEffect(useCallback(() => {
    loadTransactions();
  }, []));

  return (
    <Container>
      {isLoading ? (
        <LoadContainer>
          <ActivityIndicator color={theme.colors.secondary} size="large" />
        </LoadContainer>
      ) : (
        <>
          <Header>
            <UserWrapper>
              <UserInfo>
                <Photo source={{ uri: 'https://avatars.githubusercontent.com/u/56850413?v=4'}} />
                <User>
                  <UserGreeting>Olá,</UserGreeting>
                  <UserName>Hedênica</UserName>
                </User>
              </UserInfo>
              <LogoutButton onPress={() => {}}>
                <Icon name="power" />
              </LogoutButton>
            </UserWrapper>
          </Header>

          <CardsContainer>
            <Card
              type="income"
              title="Entradas"
              amount={cardData.income.amount}
              lastTransaction="Última entrada dia 13 de abril"
            />
            <Card
              type="outcome"
              title="Saídas"
              amount={cardData.outcome.amount}
              lastTransaction="Última saída dia 03 de abril"
            />
            <Card
              type="total"
              title="Total"
              amount={cardData.total.amount}
              lastTransaction="01 à 16 de abril"
            />
          </CardsContainer>

          <Transactions>
            <Title>Listagem</Title>

            <TransactionList 
              data={transactions}
              keyExtractor={item => item.id}
              renderItem={({ item }) => <TransactionCard data={item} />}
            />
          </Transactions>
        </>
      )}
    </Container>
  )
}
