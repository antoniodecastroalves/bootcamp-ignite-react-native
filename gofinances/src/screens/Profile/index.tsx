import React from 'react';
import { 
  View,
  Text,
  TextInput,
  Button,
} from 'react-native';

export const Profile = () => {
  return (
    <View>
      <Text testID='profile-title'>Perfil</Text>
      <TextInput 
        testID='input-name'
        placeholder='Nome'
        autoCorrect={false}
        value='HedÃªnica'
      />
      <TextInput 
        testID='input-lastname'
        placeholder='Sobrenome'
        autoCorrect={false}
        value='Morais'
      />

      <Button title='Salvar' onPress={() => {}} />
    </View>
  );
}
