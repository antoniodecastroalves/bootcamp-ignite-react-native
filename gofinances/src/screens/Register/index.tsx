import React, { useState } from 'react'
import { Keyboard, Modal, TouchableWithoutFeedback, Alert } from 'react-native'
import * as Yup from 'yup'
import { yupResolver } from '@hookform/resolvers/yup'
import AsyncStorage from '@react-native-async-storage/async-storage'
import uuid from 'react-native-uuid'

import { useForm } from 'react-hook-form'
import { useNavigation } from '@react-navigation/native'

import { InputForm } from '../../components/Form/InputForm'
import { Button } from '../../components/Form/Button'
import { TransactionTypeButton } from '../../components/Form/TransactionTypeButton'
import { CategorySelectButton } from '../../components/Form/CategorySelectButton'

import { CategorySelect } from '../CategorySelect'

import { useAuth } from '../../hooks/auth'

import {
  Container,
  Header,
  Title,
  Form,
  Fields,
  TransactionsTypes,
} from './styles'

interface FormData {
  name: string;
  amount: string;
}

type NavigationProps = {
  navigate:(screen:string) => void;
}

const schema = Yup.object().shape({
  name: Yup
    .string()
    .required('Nome obrigatório ⚠️'),
  amount: Yup
    .number()
    .typeError('Informe um valor numérico')
    .positive('O valor não poder ser negativo')
    .required('Preço obrigatório ⚠️'),
})

export function Register() {
  const {
    control,
    handleSubmit,
    reset,
    formState: { errors }
  } = useForm({ resolver: yupResolver(schema)})
  const { user } = useAuth()

  const navigation = useNavigation<NavigationProps>()

  const [categoryModalOpen, setCategoryModalOpen] = useState(false)
  const [transactionType, setTransactionType] = useState('')
  const [category, setCategory] = useState({
    key: 'category',
    name: 'Categoria',
  })

  function handleOpenCategoryModal() {
    setCategoryModalOpen(true)
  }

  function handleCloseCategoryModal() {
    setCategoryModalOpen(false)
  }

  function handleTransactionTypeSelect(type: 'income' | 'outcome') {
    setTransactionType(type)
  }

  async function handleRegister(form: FormData) {
    if (!transactionType) {
      return Alert.alert('Selecione um tipo de transação')
    }

    if (category.key === 'category') {
      return Alert.alert('Selecione uma categoria')
    }

    const newTransaction = {
      ...form,
      id: String(uuid.v4()),
      type: transactionType,
      category: category.key,
      date: new Date(),
    }

    try {
      const dataKey = `@gofinances:transactions_user:${user.id}`
      const storedData = await AsyncStorage.getItem(dataKey);
      const parsedData = storedData ? JSON.parse(storedData) : [];

      const formattedData = [...parsedData, newTransaction]

      await AsyncStorage.setItem(dataKey, JSON.stringify(formattedData))

      reset();
      setTransactionType('');
      setCategory({
        key: 'category',
        name: 'Categoria',
      });

      navigation.navigate('Listagem');

    } catch (error) {
      console.log(error)
      Alert.alert('Erro ao registrar transação')
    }
  }

  return (
    <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
      <Container>
        <Header>
          <Title>Cadastro</Title>
        </Header>
        <Form>
          <Fields>
            <InputForm
              name="name"
              control={control}
              placeholder="Nome"
              autoCapitalize="sentences"
              autoCorrect={false}
              error={errors.name && errors.name.message}
            />
            <InputForm
              name="amount"
              control={control}
              placeholder="Preço"
              keyboardType="numeric"
              error={errors.amount && errors.amount.message}
            />
            <TransactionsTypes>
              <TransactionTypeButton
                type="income"
                title="Entrada"
                isActive={transactionType === 'income'}
                onPress={() => handleTransactionTypeSelect('income')}
              />
              <TransactionTypeButton
                type="outcome"
                title="Saída"
                isActive={transactionType === 'outcome'}
                onPress={() => handleTransactionTypeSelect('outcome')}
              />
            </TransactionsTypes>
            <CategorySelectButton
              testID='button-category'
              title={category.name}
              onPress={handleOpenCategoryModal}
            />
          </Fields>
          <Button
            title="Enviar"
            onPress={handleSubmit(handleRegister)}
          />
        </Form>

        <Modal testID='modal-category' visible={categoryModalOpen}>
          <CategorySelect
            category={category}
            setCategory={setCategory}
            closeSelectCategory={handleCloseCategoryModal}
          />
        </Modal>
      </Container>
    </TouchableWithoutFeedback>
  )
}
