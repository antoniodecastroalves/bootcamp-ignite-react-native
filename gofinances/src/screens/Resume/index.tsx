import React, { useCallback, useState } from 'react'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { ActivityIndicator } from 'react-native';
import { RFValue } from 'react-native-responsive-fontsize';
import { VictoryPie } from 'victory-native'
import { addMonths, subMonths, format } from 'date-fns'
import { ptBR } from 'date-fns/locale'

import { useFocusEffect  } from '@react-navigation/native'
import { useBottomTabBarHeight } from '@react-navigation/bottom-tabs'
import { useTheme } from 'styled-components'

import { HistoryCard } from '../../components/HistoryCard'

import { categories } from '../../utils/categories';

import {
  Container,
  LoadContainer,
  Header,
  Title,
  Content,
  ChartContainer,
  MonthSelect,
  MonthSelectButton,
  MonthSelectIcon,
  Month,
} from './styles'

export interface TransactionData {
  type: 'income' | 'outcome';
  name: string;
  amount: string;
  category: string;
  date: string;
}

interface CategoryData {
  key: string;
  name: string;
  total: number;
  totalFormatted: string;
  percent: string;
  color: string;
}

export function Resume() {
  const theme = useTheme();

  const [isLoading, setLoading] = useState(false);
  const [selectedDate, setSelectedDate] = useState(new Date());
  const [totalByCategories, setTotalByCategories] = useState<CategoryData[]>([])

  function handleChangeDate(action: 'next' | 'prev') {
    if (action === 'next') {
      setSelectedDate(addMonths(selectedDate, 1))
    } else {
      setSelectedDate(subMonths(selectedDate, 1))
    }
  }

  async function getTransactions() {
    setLoading(true)
    const dataKey = '@gofinances:transactions';
    const storedData = await AsyncStorage.getItem(dataKey);
    const currentTransactions = storedData ? JSON.parse(storedData) : [];

    const outcomes = currentTransactions
      .filter((transaction: TransactionData) =>
        transaction.type === 'outcome' &&
        new Date(transaction.date).getMonth() === selectedDate.getMonth() &&
        new Date(transaction.date).getFullYear() === selectedDate.getFullYear()
      );

    const outcomesTotal = outcomes
      .reduce((accumulator: number, outcome: TransactionData) => {
        return accumulator + Number(outcome.amount);
      }, 0);

    const totalByCategory: CategoryData[] = []

    categories.forEach(category => {
      let categorySum = 0;

      outcomes.forEach((outcome: TransactionData) => {
        if (outcome.category === category.key) {
          categorySum += Number(outcome.amount);
        }
      });

      if (categorySum > 0) {
        const totalFormatted = categorySum.toLocaleString('pt-BR', {
          style: 'currency',
          currency: 'BRL',
        })

        const percent = `${(categorySum / outcomesTotal * 100).toFixed(0)}%`;

        totalByCategory.push({
          key: category.key,
          name: category.name,
          color: category.color,
          total: categorySum,
          totalFormatted,
          percent,
        })
      }

    })

    setTotalByCategories(totalByCategory);
    setLoading(false);
  }

  useFocusEffect(useCallback(() => {
    getTransactions();
  }, [selectedDate]))

  return (
    <Container>
      <Header>
        <Title>Resumo por categoria</Title>
      </Header>
      {isLoading ? (
        <LoadContainer>
          <ActivityIndicator color={theme.colors.secondary} size="large" />
        </LoadContainer>
      ) : (
        <Content
          showsVerticalScrollIndicator={false}
          contentContainerStyle={{
            paddingHorizontal: 24,
            paddingBottom: useBottomTabBarHeight(),
          }}
        >
          <MonthSelect>
            <MonthSelectButton onPress={() => handleChangeDate('prev')}>
              <MonthSelectIcon name="chevron-left" />
            </MonthSelectButton>

            <Month>
              {format(selectedDate, 'MMMM, yyyy', {
                locale: ptBR,
              })}
            </Month>

            <MonthSelectButton onPress={() => handleChangeDate('next')}>
              <MonthSelectIcon name="chevron-right" />
            </MonthSelectButton>
          </MonthSelect>

          <ChartContainer>
            <VictoryPie
              data={totalByCategories}
              x="percent"
              y="total"
              colorScale={totalByCategories.map(category => category.color)}
              style={{
                labels: {
                  fontSize: RFValue(18),
                  fontWeight: 'bold',
                  fill: theme.colors.shape,
                },
              }}
              labelRadius={60}
            />
          </ChartContainer>
          {
            totalByCategories.map(category => (
              <HistoryCard
                key={category.key}
                color={category.color}
                title={category.name}
                amount={category.totalFormatted}
              />
            ))
          }
        </Content>
      )}
    </Container>
  )
}