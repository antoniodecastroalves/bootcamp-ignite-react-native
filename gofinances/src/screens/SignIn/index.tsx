import React, { useState } from 'react'
import { ActivityIndicator, Alert, Platform } from 'react-native'
import { RFValue } from 'react-native-responsive-fontsize';
import { useTheme } from 'styled-components'

import AppleSvg from '../../assets/apple.svg'
import GoogleSvg from '../../assets/google.svg'
import LogoSvg from '../../assets/logo.svg'

import { useAuth } from '../../hooks/auth';

import { SignInSocialButton } from '../../components/SignInSocialButton';

import {
  Container,
  Header,
  TitleWrapper,
  Title,
  SignTitle,
  Footer,
  FooterWrapper,
} from './styles'

export function SignIn() {
  const [isLoading, setLoading] = useState(false);
  const { signInWithGoogle, signInWithApple } = useAuth();
  const theme = useTheme();

  function handleSignInWithGoogle() {
    try {
      setLoading(true)
      return signInWithGoogle()
    } catch (error) {
      console.log(error);
      setLoading(false)
      Alert.alert('Ops, tivemos um erro ao logar, tente novamente!')
    }
  }

  function handleSignInWithApple() {
    try {
      setLoading(true)
      return signInWithApple()
    } catch (error) {
      console.log(error);
      setLoading(false)
      Alert.alert('Ops, tivemos um erro ao logar, tente novamente!')
    }
  }

  return(
    <Container>
      <Header>
        <TitleWrapper>
          <LogoSvg
            width={RFValue(120)}
            height={RFValue(68)}
          />
          <Title>
            Controle suas {'\n'}
            finanças de forma {'\n'}
            muito simples
          </Title>
        </TitleWrapper>

        {Platform.OS === 'ios' ? (
          <SignTitle>
            Faça seu login com {'\n'}
            uma das contas abaixo
          </SignTitle>
        ) : (
          <SignTitle>
            Faça seu login com {'\n'}
            a Google
          </SignTitle>
        )}
      </Header>
      <Footer>
        <FooterWrapper>
          <SignInSocialButton
            title="Entrar com Google"
            svg={GoogleSvg}
            onPress={handleSignInWithGoogle}
          />
          {Platform.OS === 'ios' && (
            <SignInSocialButton
              title="Entrar com Apple"
              svg={AppleSvg}
              onPress={handleSignInWithApple}
            />
          )}
        </FooterWrapper>

        { isLoading && 
          <ActivityIndicator
            color={theme.colors.shape}
            style={{ marginTop: 20 }}
          />
        }
      </Footer>
    </Container>
  )
}