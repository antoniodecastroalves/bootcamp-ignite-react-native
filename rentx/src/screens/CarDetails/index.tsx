import React from 'react';
import { StatusBar, StyleSheet } from 'react-native';
import { useNavigation, useRoute } from '@react-navigation/native'
import { useTheme } from 'styled-components';

import Animated, { 
  useSharedValue,
  useAnimatedScrollHandler,
  useAnimatedStyle,
  interpolate,
  Extrapolate,
} from 'react-native-reanimated'

import { BackButton } from '../../components/BackButton';
import { ImageSlider } from '../../components/ImageSlider';
import { Accessory } from '../../components/Accessory';
import { Button } from '../../components/Button';

import { CarDTO } from '../../dtos/CarDTO';

import { getAccessoryIcon } from '../../utils/getAccessoryIcon';

import {
  Container,
  Header,
  Details,
  Description,
  Brand,
  Name,
  Rent,
  Period,
  Price,
  Accessories,
  About,
  Footer,
} from './styles';
import { getStatusBarHeight } from 'react-native-iphone-x-helper';

type NavigationProps = {
  navigate:(screen:string, car: object) => void;
  goBack: () => void;
}

interface Params {
  car: CarDTO;
}

export function CarDetails() {
  const theme = useTheme()
  const { navigate, goBack } = useNavigation<NavigationProps>();
  const route = useRoute();
  const { car } = route.params as Params;

  const scrollY = useSharedValue(0)
  const scrollHandler = useAnimatedScrollHandler(event => {
    scrollY.value = event.contentOffset.y
  })

  const headerStyleAnimation = useAnimatedStyle(() => {
    return {
      height: interpolate(
        scrollY.value,
        [0, 200],
        [200, 70],
        Extrapolate.CLAMP
      )
    }
  })

  const sliderCarsStyleAnimation = useAnimatedStyle(() => {
    return {
      opacity: interpolate(
        scrollY.value,
        [0, 150],
        [1,0],
        Extrapolate.CLAMP
      ),
    }
  })

  function handleChooseRentalPeriod() {
    navigate('Scheduling', { car });
  }

  return (
    <Container>
      <StatusBar barStyle="dark-content" />
      <Animated.View style={[
        headerStyleAnimation,
        styles.header,
        { backgroundColor: theme.colors.background_secondary }
      ]}>
        <Header>
          <BackButton onPress={() => goBack() } />
        </Header>

        <Animated.View style={[sliderCarsStyleAnimation, styles.carSlider]}>
          <ImageSlider imagesUrl={car.photos} />
        </Animated.View>
      </Animated.View>

      <Animated.ScrollView
        contentContainerStyle={{
          paddingTop: getStatusBarHeight() + 160,
          paddingHorizontal: 24,
          alignItems: 'center',
        }}
        showsVerticalScrollIndicator={false}
        onScroll={scrollHandler}
        scrollEventThrottle={16}
      >
        <Details>
          <Description>
            <Brand>{car.brand}</Brand>
            <Name>{car.name}</Name>
          </Description>
          <Rent>
            <Period>{car.period}</Period>
            <Price>R$ {car.price}</Price>
          </Rent>
        </Details>
        <Accessories>
          {car.accessories.map(accessory => (
            <Accessory
              key={accessory.type}
              name={accessory.name}
              icon={getAccessoryIcon(accessory.type)} 
            />
          ))}
        </Accessories>
        <About>
          {car.about}
          {car.about}
          {car.about}
          {car.about}
          {car.about}
        </About>
      </Animated.ScrollView>

      <Footer>
        <Button title="Escolher perÃ­odo do aluguel" onPress={handleChooseRentalPeriod} />
      </Footer>

    </Container>
  );
}

const styles = StyleSheet.create({
  header: {
    position: 'absolute',
    overflow: 'hidden',
    zIndex: 1,
  },
  carSlider: {
    marginTop: getStatusBarHeight() + 32
  }
})