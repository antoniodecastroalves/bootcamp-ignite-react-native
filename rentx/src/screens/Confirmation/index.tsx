import React from 'react';
import { StatusBar, useWindowDimensions } from 'react-native'
import { useNavigation, useRoute } from '@react-navigation/native'

import LogoSvg from '../../assets/logo_background_gray.svg';
import DoneSvg from '../../assets/done.svg';

import { ConfirmButton } from '../../components/ConfirmButton';

import {
  Container,
  Content,
  Title,
  Message,
  Footer,
} from './styles';

type NavigationProps = {
  navigate:(screen:string) => void;
}

type Params = {
  title: string
  message: string
  nextScreen: string
}

export function Confirmation() {
  const { width } = useWindowDimensions()
  const { navigate } = useNavigation<NavigationProps>();
  const { params } = useRoute()
  const { title, message, nextScreen } = params as Params

  function handleReturnHome() {
    navigate(nextScreen);
  }

  return (
    <Container>
      <StatusBar barStyle="light-content" backgroundColor="transparent" translucent />
      <LogoSvg width={width} />

      <Content>
        <DoneSvg width={80} height={80} />
        <Title>{title}</Title>
        <Message>{message}</Message>
      </Content>
      <Footer>
        <ConfirmButton title="ok" onPress={handleReturnHome} />
      </Footer>
    </Container>
  );
}