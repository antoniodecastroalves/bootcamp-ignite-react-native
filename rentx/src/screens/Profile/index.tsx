import React, { useState } from 'react';
import { KeyboardAvoidingView, TouchableWithoutFeedback, Keyboard, Alert } from 'react-native';
import { useTheme } from 'styled-components';
import { BackButton } from '../../components/BackButton';
import { useNavigation } from '@react-navigation/native';
import { useBottomTabBarHeight } from '@react-navigation/bottom-tabs';
import { Feather } from '@expo/vector-icons';
import * as ImagePicker from 'expo-image-picker';
import { ImageInfo } from 'expo-image-picker/build/ImagePicker.types';
import * as Yup from 'yup';

import {
  Container,
  Header,
  HeaderTop,
  HeaderTitle,
  SignOutButton,
  PhotoContainer,
  Photo,
  PhotoButton,
  Content,
  Options,
  Option,
  OptionTitle,
  Section,
} from './styles';
import { Button } from '../../components/Button';
import { Input } from '../../components/Input';
import { PasswordInput } from '../../components/PasswordInput/index';
import { useAuth } from '../../hooks/auth';

type Options = 'dataEdit' | 'passwordEdit'

export function Profile() {
  const theme = useTheme();
  const navigation = useNavigation();
  const { user, signOut, updateUser } = useAuth()

  const [option, setOption] = useState<Options>('dataEdit');
  const [avatar, setAvatar] = useState(user.avatar);
  const [name, setName] = useState(user.name);
  const [driverLicense, setDriverLicense] = useState(user.driver_license);

  const handleBack = () => navigation.goBack()

  const handleOptionPress = (selectedOption: Options) => setOption(selectedOption)

  const handleChangeAvatar = async () => {
    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [4, 4],
      quality: 1,
    })

    const { uri } = result as ImageInfo

    if (result.cancelled) { 
      return 
    }

    if (uri) {
      setAvatar(uri)
    }
  }

  const handleUserUpdate = async () => {
    try {
      const schema = Yup.object().shape({
        driverLicense: Yup.string().required('CNH é obrigatório'),
        name: Yup.string().required('Nome é obrigatório')
      })

      const data = { name, driverLicense }

      await schema.validate(data);

      await updateUser({
        ...user,
        name,
        driver_license: driverLicense,
        avatar,
      });

      Alert.alert('Perfil atualizado com sucesso ✅')

    } catch (error) {
      if (error instanceof Yup.ValidationError) {
       Alert.alert('Opa', error.message)
      }
      console.log(error)
      Alert.alert('Não foi possível atualizar o perfil, tente novamente.')
    }
  }

  const handleSignOut = () => {
    Alert.alert(
      'Tem certeza? 👀',
      'Lembre-se, que precisará de internet para logar novamente.',
      [
        { text: 'Cancelar ❌', onPress: () => {} },
        { text: 'Sair ✅', onPress: () => signOut() }
      ]
    )
  }

  return (
    <KeyboardAvoidingView behavior='position' enabled>
      <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
        <Container>
          <Header>
            <HeaderTop>
              <BackButton color={theme.colors.shape} onPress={handleBack} />
              <HeaderTitle>Editar Perfil</HeaderTitle>
              <SignOutButton onPress={handleSignOut}>
                <Feather name='power' size={24} color={theme.colors.shape} />
              </SignOutButton>
            </HeaderTop>
            <PhotoContainer>
              { !!avatar && <Photo source={{ uri: avatar }} />}
              <PhotoButton onPress={handleChangeAvatar}>
                <Feather name='camera' size={24} color={theme.colors.shape} />
              </PhotoButton>
            </PhotoContainer>
          </Header>

          <Content style={{ marginBottom: useBottomTabBarHeight()}}>
            <Options>
              <Option 
                active={option === 'dataEdit'}
                onPress={() => handleOptionPress('dataEdit')}
              >
                <OptionTitle active={option === 'dataEdit'}>
                  Dados
                </OptionTitle>
              </Option>
              <Option 
                active={option === 'passwordEdit'}
                onPress={() => handleOptionPress('passwordEdit')}
              >
                <OptionTitle active={option === 'passwordEdit'}>
                  Trocar senha
                </OptionTitle>
              </Option>
            </Options>
            {option === 'dataEdit' ? (
              <Section>
                <Input 
                  iconName='user'
                  placeholder='Nome'
                  autoCorrect={false}
                  defaultValue={user.name}
                  onChangeText={setName}
                />
                <Input 
                  iconName='mail'
                  editable={false}
                  defaultValue={user.email}
                />
                <Input 
                  iconName='credit-card'
                  placeholder='CNH'
                  keyboardType='numeric'
                  defaultValue={user.driver_license}
                  onChangeText={setDriverLicense}
                />
              </Section>
            ): (
              <Section>
                <PasswordInput iconName='lock' placeholder='Senha atual' />
                <PasswordInput iconName='lock' placeholder='Nova senha' />
                <PasswordInput iconName='lock' placeholder='Repetir senha' />
              </Section>
            )}
            <Button 
              title='Salvar alterações'
              onPress={handleUserUpdate}
            />
          </Content>
        </Container>
      </TouchableWithoutFeedback>
    </KeyboardAvoidingView>
  );
}