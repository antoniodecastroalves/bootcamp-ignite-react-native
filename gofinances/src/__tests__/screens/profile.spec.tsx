import React from 'react';
import { render } from '@testing-library/react-native'

import { Profile } from '../../screens/Profile';

describe('Profile Screen', () => {
  it('renders correctly', () => {
    const component = render(<Profile />)

    expect(component).toMatchSnapshot()
  })

  it('renders correctly the user input name placeholder', () => {
    const { getByPlaceholderText } = render(<Profile />)

    const placeholder = getByPlaceholderText('Nome')

    expect(placeholder).toBeTruthy()
  })

  it('loads the user data', () => {
    const { getByTestId } = render(<Profile />)

    const inputName = getByTestId('input-name')
    const inputLastname = getByTestId('input-lastname')

    expect(inputName.props.value).toEqual('HedÃªnica')
    expect(inputLastname.props.value).toEqual('Morais')

  })

  it('render the title correctly', () => {
    const { getByTestId } = render(<Profile />)

    const textTitle = getByTestId('profile-title')

    expect(textTitle.props.children).toContain('Perfil')
  })
});