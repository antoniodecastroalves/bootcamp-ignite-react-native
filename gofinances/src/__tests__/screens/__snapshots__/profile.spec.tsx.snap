// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Screen renders correctly 1`] = `
<View>
  <Text
    testID="profile-title"
  >
    Perfil
  </Text>
  <TextInput
    allowFontScaling={true}
    autoCorrect={false}
    placeholder="Nome"
    rejectResponderTermination={true}
    testID="input-name"
    underlineColorAndroid="transparent"
    value="HedÃªnica"
  />
  <TextInput
    allowFontScaling={true}
    autoCorrect={false}
    placeholder="Sobrenome"
    rejectResponderTermination={true}
    testID="input-lastname"
    underlineColorAndroid="transparent"
    value="Morais"
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Salvar
      </Text>
    </View>
  </View>
</View>
`;
